cpu/sam0_common/sam0_eth/eth-netdev.c

```diff
@@ -205,10 +205,35 @@ static int _sam0_eth_send(netdev_t *netdev, const iolist_t *iolist)
         /* TODO: use a specific netdev callback here ? */
         return -EOVERFLOW;
     }
-    netdev->event_callback(netdev, NETDEV_EVENT_TX_COMPLETE);
     return ret;
 }
 
+static int _sam0_eth_confirm_send(netdev_t *netdev, void *info)
+{
+    (void)netdev;
+    (void)info;
+
+    uint32_t tsr = GMAC->TSR.reg;
+    GMAC->TSR.reg = tsr; /* clear flags */
+
+    /* transmit is active */
+    if (tsr & GMAC_TSR_TXGO) {
+        return -EAGAIN;
+    }
+
+    /* Retry Limit Exceeded */
+    if (tsr & GMAC_TSR_RLE) {
+        return -EBUSY;
+    }
+
+    /* Transmit Frame Corruption, Collision Occurred, Transmit Underrun */
+    if (tsr & (GMAC_TSR_TFC | GMAC_TSR_COL | GMAC_TSR_UND)) {
+        return -EIO;
+    }
+
+    return 1;
```

Comment:
Ideally it should return the bytes actually transmitted instead of `1` here for `netstats` to be correct.

Reply:
Makes me wonder: what happens with the return value of `.send()`?

Reply:
The return valur of `.send()` should be `0` on "no error as of now". This doesn't rule out that confirm_send reports an error later on.

The reasoning is that crazy things that change the size of the frame like bit-stuffing, escaping, padding (or whatever may apply to the netdev) can be done during TX, and is not needed to be accounted for when accepting the frame. After the fact it might be cheaper to provide the actual byte count.

https://doc.riot-os.org/structnetdev__driver.html#a39585137953a92c1c58b5e0ad3262096

Reply:
Ah good catch, I'm still returning the size of the frame as with the old API